### PR TITLE
net: dns: Add DNS cache for improved performance

### PR DESCRIPTION
**Pull request by kica-z**
In order to reduce dns resolve requests when using the dns resolver i added a caching mechanism. I would appreciate some feedback as to if a cache is even something we should add as well as opinions regarding implementation. I have described the current implementation as well as known problems here.

**Implementation:**
The current implementation is as follows:
- One cache for all DNS resolver contexts
- The cache returns as many results as possible for each query
- Multiple queries per entry are supported
- Entries are retained based on their RR TTL and removed as soon as it runs out
- If the cache is full the oldest entry is replaced with the new addition

**Known limitations:**
- When the resolver supports multiple addrinfo per query (CONFIG_DNS_RESOLVER_AI_MAX_ENTRIES) and those entries have different TTLs the cache will at some point no longer return all entries due to TTL expiry. This could be the expected behaviour but also might not be what the user wants.
- If multiple DNS resolver contexts run the same query in parallel duplicate entries could be introduced into the cache since the same address will be added twice.

### CODE REVIEW

File: tests/net/lib/dns_cache/src/main.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+void clear_cache(void)
+{
+	dns_cache_flush(&test_dns_cache);
+}
+
+ZTEST_SUITE(net_dns_cache_test, NULL, NULL, clear_cache, NULL, NULL);
+
+ZTEST(net_dns_cache_test, test_simple_cache_entry)
+{
+	struct dns_addrinfo info_write = {.ai_family = 42};
```
Comment (jukkar) [inline tests/net/lib/dns_cache/src/main.c:25]: Is the 42 value here just some arbitrary value, why not use proper address family values?

File: include/zephyr/net/dns_resolve.h [hunk 2]
```diff
@@ -82,6 +82,13 @@ enum dns_query_type {
 
 #define DNS_MAX_MCAST_SERVERS (MDNS_SERVER_COUNT + LLMNR_SERVER_COUNT)
 
+/* RFC 1035, 3.1. Name space definitions
+ * To simplify implementations, the total length of a domain name (i.e.,
+ * label octets and label length octets) is restricted to 255 octets or
+ * less.
+ */
+#define DNS_MAX_NAME_LEN	255
```
[the remark this answers is not in the document]
Comment (kica-z) [reply include/zephyr/net/dns_resolve.h:90]: A Kconfig option would surely be a good solution however this was already defined within `dns_resolve.c` thats why I reused it. Do you think this can be replaced in both files with this new kconfig option?

File: include/zephyr/net/dns_resolve.h [hunk 3]
```diff
@@ -82,6 +82,13 @@ enum dns_query_type {
```
[the remark this answers is not in the document]
Comment (kica-z) [reply include/zephyr/net/dns_resolve.h:82]: Done 👍

File: subsys/net/lib/dns/Kconfig [hunk 4]
```diff
@@ -128,6 +128,27 @@ module-str = Log level for DNS resolver
 module-help = Enables DNS resolver code to output debug messages.
 source "subsys/net/Kconfig.template.log_config.net"
 
+menuconfig DNS_RESOLVER_CACHE
+	bool "DNS resolver cache"
+	default n
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/Kconfig]: Done 👍

File: subsys/net/lib/dns/dns_cache.c [hunk 5]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2024 Endress+Hauser AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/dns_resolve.h>
+#include "dns_cache.h"
+
+LOG_MODULE_REGISTER(net_dns_cache, 4);
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Thanks for pointing this out. I must have forgotten to adjust this after debugging.

File: subsys/net/lib/dns/dns_cache.c [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	int64_t oldest_entry_uptime = INT64_MAX;
+	size_t index_to_replace = 0;
+	bool found_empty = false;
+
+	if (cache == NULL || query == NULL || addrinfo == NULL || ttl == 0) {
+		return -EINVAL;
+	}
+
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		LOG_WRN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: I replaced the normal log macros. A question out of interest, what is the difference? Both only show when NET_LOG is activated since `CONFIG_DNS_RESOLVER_LOG_LEVEL` only gets set when this is the case. So I am wondering if there is another difference. Sorry if this might be a stupid question.

File: subsys/net/lib/dns/resolve.c [hunk 7]
```diff
@@ -1184,6 +1186,23 @@ int dns_resolve_name(struct dns_resolve_context *ctx,
...
+		/* The query was cached, no
+		 * need to continue further.
+		 */
+		for (size_t cache_index = 0; cache_index < ret; cache_index++) {
+			cb(DNS_EAI_INPROGRESS, &cached_info[cache_index], user_data);
+		}
+		cb(DNS_EAI_ALLDONE, NULL, user_data);
+
+		return 0;
+	}
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/resolve.c:1203]: I totally agree with you that this is a very simple implementation. Also as mentioned in the PR notes there are already some known limitations. That is the reason why I would appreciate some feedback on this in order to evaluate what the needs are and how much additional complexity we need.

File: tests/net/lib/dns_cache/src/main.c (hunk 1 above)
Comment (kica-z) [reply tests/net/lib/dns_cache/src/main.c]: It is not arbitrary (https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy#Answer_to_the_Ultimate_Question_of_Life,_the_Universe,_and_Everything_(42)) but definitely only a sentinel value 😅.

I have replaced it with a proper address family.

File: include/zephyr/net/dns_resolve.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply include/zephyr/net/dns_resolve.h:82]: This one PR is good, no need to have multiple PRs.

File: subsys/net/lib/dns/dns_cache.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/lib/dns/dns_cache.c]: The NET_LOG macros are mostly same as normal logging macros, but for example NET_DBG prints the thread name or id which normal logging macro does not by default. The idea with separate networking logging macros is that we can adjust the printouts more easily for network specific stuff if needed.

File: include/zephyr/net/dns_resolve.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply include/zephyr/net/dns_resolve.h:90]: Hmm, that is a good question. I am a bit worried that the value might be too large for normal use and users might want to tweak it to save memory. So if we have a Kconfig option it should be used in all over the place.
@rlubos what is you opinion about this?

File: subsys/net/lib/dns/dns_cache.c [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	int64_t oldest_entry_uptime = INT64_MAX;
+	size_t index_to_replace = 0;
+	bool found_empty = false;
+
+	if (cache == NULL || query == NULL || addrinfo == NULL || ttl == 0) {
+		return -EINVAL;
+	}
+
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
Comment (jukkar) [inline subsys/net/lib/dns/dns_cache.c:37]: If we set the max length behind the kconfig option, then we need to print the value here instead of the name.

File: subsys/net/lib/dns/dns_cache.c [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	k_mutex_unlock(cache->lock);
+
+	return 0;
+}
+
+int dns_cache_remove(struct dns_cache *cache, char const *query)
+{
+	NET_DBG("REMOVE all entries with query \"%s\"", query);
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
Comment (jukkar) [inline subsys/net/lib/dns/dns_cache.c:79]: ditto

File: subsys/net/lib/dns/dns_cache.c [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		   size_t addrinfo_array_len)
+{
+	size_t found = 0;
+
+	NET_DBG("FIND \"%s\"", query);
+	if (cache == NULL || query == NULL || addrinfo == NULL || addrinfo_array_len <= 0) {
+		return -EINVAL;
+	}
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
Comment (jukkar) [inline subsys/net/lib/dns/dns_cache.c:109]: and here too

File: subsys/net/lib/dns/dns_cache.h [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+ * @brief Statically define and initialize a DNS queue.
+ *
+ * The cache can be accessed outside the module where it is defined using:
+ *
+ * @code extern struct dns_cache <name>; @endcode
+ *
+ * @param name Name of the cache.
+ */
+#define DNS_CACHE_DEFINE(name, cache_size)                                                         \
+	K_MUTEX_DEFINE(name##_mutex);                                                              \
```
Comment (jukkar) [inline subsys/net/lib/dns/dns_cache.h:44]: This could be static too

File: subsys/net/lib/dns/dns_cache.h [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+	static struct dns_cache_entry name##_entries[cache_size] = {0};                            \
```
Comment (jukkar) [inline subsys/net/lib/dns/dns_cache.h:45]: No need to initialize to 0 because this is a static variable.

File: subsys/net/lib/dns/resolve.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/lib/dns/resolve.c:1203]: Ok, we can go on with this minimalistic approach first and see if anything else needs to be added. One important thing that I think we need to add with this first version is negative caching, meaning adding information that some dns name was not found.

File: subsys/net/lib/dns/dns_cache.c [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+		LOG_WRN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
+			strlen(query));
+		return -EINVAL;
+	}
+
+	k_mutex_lock(cache->mut, K_FOREVER);
+
+	LOG_DBG("ADD \"%s\" with TTL %" PRIu32, query, ttl);
+
+	dns_cache_clean(cache);
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c:46]: True but then we would need a loop more as soon as the cache is full every time we add something (without calling find). Do you think this would still be more efficient? Or is there something I am missing here?
```c
        // Additional loop to check if space is available
	for (size_t i = 0; i < cache->size; i++) {
		if (!cache->entries[i].in_use) {
			index_to_replace = i;
			found_empty = true;
		}
	}

	if(!found_empty) {
		dns_cache_clean(cache);
		for (size_t i = 0; i < cache->size; i++) {
			if (!cache->entries[i].in_use) {
				index_to_replace = i;
				found_empty = true;
				break;
			} else if (cache->entries[i].creation_uptime < oldest_entry_uptime) {
				index_to_replace = i;
				oldest_entry_uptime = cache->entries[i].creation_uptime;
			}
		}
	}
```

File: subsys/net/lib/dns/dns_cache.h (hunk 11 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.h]: Done

File: subsys/net/lib/dns/dns_cache.h (hunk 12 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.h]: Done

File: subsys/net/lib/dns/dns_cache.c (hunk 10 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Adjusted

File: subsys/net/lib/dns/dns_cache.c (hunk 9 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Adjusted

File: subsys/net/lib/dns/dns_cache.c (hunk 8 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Adjusted

File: include/zephyr/net/dns_resolve.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kica-z) [reply include/zephyr/net/dns_resolve.h]: @jukkar I added the Kconfig option for now it would not be a big change to revert this.

File: subsys/net/lib/dns/resolve.c [hunk 14]
```diff
@@ -635,6 +633,10 @@ int dns_validate_msg(struct dns_resolve_context *ctx,
 
 			invoke_query_callback(DNS_EAI_INPROGRESS, &info,
 					      &ctx->queries[*query_idx]);
+#ifdef CONFIG_DNS_RESOLVER_CACHE
+			dns_cache_add(&dns_cache,
```
[the remark this answers is not in the document]
Comment (kica-z) [reply subsys/net/lib/dns/resolve.c]: I agree with you that negativ caching would be nice. The question is how to go about this. Normally the TTL for negative caching entries would be set to the value of the `MINIMUM` field of the `SOA`. Currently as far as i know we do not fetch the `SOA` so do you think we should just set a sensible value for the TTL or is it worth it to fetch the `SOA` in order to add a negative cache entry?

File: subsys/net/lib/dns/resolve.c [hunk 15]
```diff
@@ -464,14 +440,12 @@ static inline int get_slot_by_id(struct dns_resolve_context *ctx,
...
 static
 #endif
-int dns_validate_msg(struct dns_resolve_context *ctx,
-		     struct dns_msg_t *dns_msg,
-		     uint16_t *dns_id,
-		     int *query_idx,
-		     struct net_buf *dns_cname,
-		     uint16_t *query_hash)
+	int
+	dns_validate_msg(struct dns_resolve_context *ctx, struct dns_msg_t *dns_msg,
```
Comment (rlubos) [inline subsys/net/lib/dns/resolve.c:444]: This indentation looks weird, why is the type indented? Anyway, a general note, it's a bit hard to review due to all those formatting changes, hard to spot the actual meaningful ones. Can we drop it from the PR?

File: tests/net/lib/dns_cache/on_target.conf [hunk 16]
```diff
@@ -0,0 +1,15 @@
+CONFIG_ZTEST=y
+
+CONFIG_NETWORKING=y
+CONFIG_NET_TEST=y
+CONFIG_NET_L2_ETHERNET=n
+
+CONFIG_MAIN_STACK_SIZE=1344
+CONFIG_DNS_RESOLVER=y
+CONFIG_DNS_RESOLVER_CACHE=y
+CONFIG_LOG=y
```
Comment (rlubos) [inline tests/net/lib/dns_cache/on_target.conf:10]: Do we need logs enabled in tests by default? It increases the image size and is not really needed when executing tests. I recall cases from the past, that tests would not fit on some platforms due to logging enabled.

File: tests/net/lib/dns_cache/src/main.c [hunk 17]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2024 Endress+Hauser AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+#include <zephyr/logging/log.h>
+#include "dns_cache.h"
+LOG_MODULE_REGISTER(net_dns_cache_test, CONFIG_DNS_RESOLVER_LOG_LEVEL);
```
Comment (rlubos) [inline tests/net/lib/dns_cache/src/main.c:10]: It doesn't seem that logging is used anywhere in the test, please remove it.

File: tests/net/lib/dns_cache/prj.conf [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+CONFIG_MAIN_STACK_SIZE=1344
+CONFIG_DNS_RESOLVER=y
+CONFIG_DNS_RESOLVER_CACHE=y
+CONFIG_LOG=y
+CONFIG_NET_LOG=y
+CONFIG_DNS_RESOLVER_LOG_LEVEL_DBG=y
+
+CONFIG_UBSAN=y
+CONFIG_ASAN=y
```
Comment (rlubos) [inline tests/net/lib/dns_cache/prj.conf:15]: IMO default `prj.conf` shoud be generic, POSIX-specifc configs should either go in a custom config file, or even better, as  `extra_configs` for posix target in `testcase.yml`.

File: subsys/net/lib/dns/dns_cache.c [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	k_mutex_unlock(cache->lock);
+
+	if (found > addrinfo_array_len) {
+		return -ENOSR;
+	}
+
+	if (found == 0) {
+		NET_DBG("COULD NOT FIND \"%s\"", query);
```
Comment (rlubos) [inline subsys/net/lib/dns/dns_cache.c:154]: Please don't use all caps in logs.

File: subsys/net/lib/dns/dns_cache.c [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+
+	for (size_t i = 0; i < cache->size; i++) {
+		if (!cache->entries[i].in_use) {
+			index_to_replace = i;
+			found_empty = true;
+		}
+	}
+
+	if (!found_empty) {
+		dns_cache_clean(cache);
```
Comment (rlubos) [inline subsys/net/lib/dns/dns_cache.c:55]: Maybe just clean the cache before the first loop, so that we don't need to check for `in_use` twice?

File: subsys/net/lib/dns/dns_cache.c [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+/* Needs to be called when lock is already acquired */
+static void dns_cache_clean(struct dns_cache const *cache)
+{
+	for (size_t i = 0; i < cache->size; i++) {
+		if (!cache->entries[i].in_use) {
+			continue;
+		}
+		int64_t reftime = cache->entries[i].creation_uptime;
```
Comment (rlubos) [inline subsys/net/lib/dns/dns_cache.c:166]: About the time calculations - I wonder that whether it'd make sense, instead of storing creation time + ttl, just have a `k_timepoint_t` for the expiry time, then you could simply use `sys_timepoint_expired()` to check if the entry is still valid, instead of doing the calculations manually.

File: subsys/net/lib/dns/resolve.c (hunk 15 above)
Comment (kica-z) [reply subsys/net/lib/dns/resolve.c]: Fixed. Removed formatting changes.

File: tests/net/lib/dns_cache/on_target.conf (hunk 16 above)
Comment (kica-z) [reply tests/net/lib/dns_cache/on_target.conf]: Removed log kconfigs

File: subsys/net/lib/dns/dns_cache.c (hunk 19 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Adjusted

File: subsys/net/lib/dns/dns_cache.c (hunk 20 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: Thats what I had before. Reverted to stage where this was the case.

File: subsys/net/lib/dns/dns_cache.c (hunk 21 above)
Comment (kica-z) [reply subsys/net/lib/dns/dns_cache.c]: This did however change the functionality a bit since now the entry closest to expiry is the one that will be removed instead of the oldest when additional space is needed.

File: include/zephyr/net/dns_resolve.h [hunk 22]
```diff
@@ -89,49 +89,49 @@ enum dns_query_type {
  */
 struct dns_addrinfo {
 	struct sockaddr ai_addr;
-	socklen_t       ai_addrlen;
-	uint8_t            ai_family;
-	char            ai_canonname[DNS_MAX_NAME_SIZE + 1];
+	socklen_t ai_addrlen;
+	uint8_t ai_family;
+	char ai_canonname[DNS_MAX_NAME_SIZE + 1];
```
Comment (jukkar) [inline include/zephyr/net/dns_resolve.h:94]: This seems to be unrelated change, please remove this, or add this to new commit if you think this should be there.

File: include/zephyr/net/dns_resolve.h [hunk 23]
```diff
@@ -89,49 +89,49 @@ enum dns_query_type {
...
 };
 
 /**
  * Status values for the callback.
  */
 enum dns_resolve_status {
 	/** Invalid value for `ai_flags' field */
-	DNS_EAI_BADFLAGS    = -1,
```
Comment (jukkar) [inline include/zephyr/net/dns_resolve.h:102]: Same for these cosmetic changes.

File: include/zephyr/net/dns_resolve.h [hunk 24]
```diff
@@ -175,10 +174,10 @@ struct dns_resolve_context {
 		struct net_context *net_ctx;
 
 		/** Is this server mDNS one */
-		uint8_t is_mdns : 1;
+		uint8_t is_mdns: 1;
 
 		/** Is this server LLMNR one */
-		uint8_t is_llmnr : 1;
+		uint8_t is_llmnr: 1;
```
Comment (jukkar) [inline include/zephyr/net/dns_resolve.h:180]: and here too. In fact it seems that all the changes to `dns_resolve.h` are cosmetic and should be removed from this PR.

File: tests/net/lib/dns_cache/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jukkar) [inline tests/net/lib/dns_cache/CMakeLists.txt:1]: Please change the subject of test commit to `tests: net: dns: ....`

**Review by jukkar — COMMENTED** [review]
Looks pretty good, please check the test failure in CI.

**Comment by kica-z** [comment]
@jukkar I addressed your newest comments (still investigating regarding CI).
Regarding negative caching should we implement this now or later? I mentioned in my https://github.com/zephyrproject-rtos/zephyr/pull/68825#discussion_r1485928466 that the `SOA` ist still unclear.

**Comment by jukkar** [comment]
> Regarding negative caching should we implement this now or later?

We can do it later too.

**Comment by kica-z** [comment]
@jukkar Build error has been fixed. Let me know if something still needs to be done.

File: tests/net/lib/dns_cache/src/main.c [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+		"Cache entry adding should work.");
+	zassert_equal(3, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(2, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(1, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(1));
```
[the remark this answers is not in the document]
Comment (kica-z) [reply tests/net/lib/dns_cache/src/main.c:132]: @jukkar 

So that some time can pass between the two calls to `dns_cache_find` and the found entry count should still remain the same since not enough time has passed.

I can remove it if you want in this particular case i don't think it makes much of a difference its only semantically more correct in my opinion.

[the remark this answers is not in the document]
Comment (jukkar) [reply tests/net/lib/dns_cache/src/main.c:132]: These sleeps seem to wait for expiration, so that looks like something we cannot avoid. Could we make the sleeps smaller, like <1sec?

[the remark this answers is not in the document]
Comment (kica-z) [reply tests/net/lib/dns_cache/src/main.c:132]: The problem is, that the TTL in DNS is expressed in seconds. We could internally store it in milliseconds and then adjust the test accordingly. Do you think this would be a good idea? It is an antipattern to adjust code to better fit the tests but there would be a significant test time decrease.

Alternatively is there a way to speed up time somehow?

[the remark this answers is not in the document]
Comment (rlubos) [reply tests/net/lib/dns_cache/src/main.c:132]: We could limit the testing scope to `native_sim` platform only, which does not execute tests in real time (with `integration_platforms` option). Thoughts?

[the remark this answers is not in the document]
Comment (kica-z) [reply tests/net/lib/dns_cache/src/main.c:132]: Sounds good to me 👍 

@jukkar @rlubos  I have adjusted the `testcase.yaml` accordingly and the checks are green.

**Review by jukkar — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
